File: src/freq01.scala
import java.io._
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.charset.StandardCharsets
import java.nio.file.Paths
import java.util

import cats.data.{ NonEmptyList, ValidatedNel }
import cats.effect._
import cats.implicits._
import com.monovore.decline._
import com.monovore.decline.effect._
import fs2._
import it.unimi.dsi.fastutil.bytes.ByteArrayList
import it.unimi.dsi.fastutil.objects.ObjectArrays

import scala.annotation.tailrec

package freq01 {
  object App extends CommandIOApp("freq", "Counts '[a-zA-Z]+' words in input", version = "0.1.0") {
    private val a = 'a'.toByte
    private val z = 'z'.toByte
    private val A = 'A'.toByte
    private val Z = 'Z'.toByte

    def main: Opts[IO[ExitCode]] = Args.parse.map { args =>
      Blocker[IO]
        .use { blocker =>
          args
            .input[IO](blocker)
            .through(collect)
            .flatMap { dict =>
              Stream.fromIterator[IO](dict.drain)
            }
            .map {
              case (counter, word) => f"$counter%d $word%s%n"
            }
            .through(args.output(blocker))
            .compile
            .drain
        }
        .as(ExitCode.Success)
    }

    def collect[F[_]]: Pipe[F, ByteBuffer, FrequencyDict] = { buffers =>
      def loop(
          s: Stream[F, ByteBuffer],
          dict: FrequencyDict,
          lastHash: Int,
          lastWord: ByteArrayList
      ): Pull[F, FrequencyDict, Unit] =
        s.pull.uncons1.flatMap {
          case None =>
            if (lastWord.isEmpty) Pull.output1(dict)
            else Pull.output1(dict.register(lastHash, lastWord.toArray(new Array[Byte](lastWord.size()))))
          case Some(buffer -> nxt) =>
            var hash = lastHash
            val word = lastWord

            while (buffer.remaining() > 0) {
              var byte = buffer.get()
              if (a <= byte && byte <= z) {
                hash = Fnv1.next(hash, byte)
                word.add(byte)
              } else if (A <= byte && byte <= Z) {
                byte = (byte | 0x20).toByte
                hash = Fnv1.next(hash, byte)
                word.add(byte)
              } else if (!word.isEmpty) {
                dict.register(hash, word.toArray(new Array[Byte](word.size())))
                word.clear()
                hash = Fnv1.H
              }
            }

            loop(nxt, dict, hash, word)
        }

      loop(buffers, FrequencyDict(), Fnv1.H, new ByteArrayList(256)).stream
    }

  }

  final case class Args(in: Option[File], out: Option[File], chunkSize: Int, bufferSize: Option[Int]) {
    def input[F[_]](blocker: Blocker)(implicit F: Sync[F], CS: ContextShift[F]): Stream[F, ByteBuffer] =
      in.map { file =>
          Stream
            .bracket {
              blocker.delay(new FileInputStream(file).getChannel)
            } { channel =>
              blocker.delay(channel.close())
            }
            .flatMap { channel =>
              Stream.unfoldEval(0L -> math.min(channel.size(), Int.MaxValue.toLong)) {
                case (_, 0L) => F.pure(none[(ByteBuffer, (Long, Long))])
                case (p, sz) =>
                  val size = math.min(sz, Int.MaxValue.toLong)
                  blocker
                    .delay(channel.map(FileChannel.MapMode.READ_ONLY, p, size))
                    .widen[ByteBuffer]
                    .tupleRight((p + size) -> (channel.size() - p - size))
                    .map(_.some)
              }
            }
        }
        .getOrElse {
          io.stdin(chunkSize, blocker).chunks.map(_.toByteBuffer)
        }

    def output[F[_]](blocker: Blocker)(implicit F: Sync[F], CS: ContextShift[F]): Pipe[F, String, Unit] = { lines =>
      val bytes = lines.flatMap { str =>
        Stream.chunk(Chunk.array(str.getBytes(StandardCharsets.UTF_8)))
      }
      val writer = out
        .map { file =>
          val fos = F
            .catchNonFatal {
              new BufferedOutputStream(new FileOutputStream(file, false))
            }
            .widen[OutputStream]
          io.writeOutputStream(fos, blocker)
        }
        .getOrElse {
          io.stdout(blocker)
        }
      bytes.through(writer)
    }

  }

  object Args {
    def stringToFile(fileName: String): ValidatedNel[String, File] =
      Either
        .catchNonFatal {
          Paths.get(fileName).toFile
        }
        .toValidated
        .leftMap(e => s"Unable to open '$fileName': ${e.getMessage}")
        .toValidatedNel

    val in: Opts[Option[File]] = Opts
      .argument[String]("input.txt")
      .mapValidated {
        case "-" => none.validNel[String]
        case fnm =>
          stringToFile(fnm)
            .ensure(NonEmptyList.one(s"'$fnm' not exists"))(_.exists())
            .ensure(NonEmptyList.one(s"'$fnm' is not a file"))(_.isFile())
            .ensure(NonEmptyList.one(s"'$fnm' can not be read"))(_.canRead())
            .map(_.some)
      }

    val out: Opts[Option[File]] = Opts
      .argument[String]("output.txt")
      .mapValidated {
        case "-" => none.validNel[String]
        case fnm =>
          stringToFile(fnm)
            .ensure(NonEmptyList.one(s"'$fnm' can not be overwritten")) { f =>
              !f.exists() || f.canWrite
            }
            .map(_.some)
      }
      .withDefault(none)

    val bufferSize: Opts[Option[Int]] = Opts
      .option[Int]("buffer-size", "Length of read buffer.", short = "b")
      .orNone

    val chunkSize: Opts[Int] = Opts
      .option[Int]("chunk-size", "Length of one chunk when processing.", short = "c")
      .withDefault(16 * 1024)

    val parse: Opts[Args] = (in, out, chunkSize, bufferSize).mapN(Args(_, _, _, _))
  }

  object Fnv1 {
    val H: Int = 0x811c9dc5
    val P: Int = 0x01000193

    def next(hash: Int, value: Byte): Int = (hash ^ value) * P
  }

  final class FrequencyDict(initial: Int) {
    import FrequencyDict.{ LoadFactor, Value }

    private var capacity = initial
    private var length   = 0
    private var mask     = capacity - 1
    private var max      = (LoadFactor * capacity.toFloat).toInt

    private var hashes = new Array[Int](capacity)
    private var values = new Array[Value](capacity)

    def size: Int = length

    def drain: Iterator[(Int, String)] = {
      val data = values

      hashes = new Array[Int](initial)
      values = new Array[Value](initial)

      capacity = initial
      length = 0
      mask = capacity - 1
      max = (LoadFactor * capacity.toFloat).toInt

      // scalafix:off DisableSyntax.null; keeping buckets sparse for locality
      ObjectArrays.unstableSort(data, { (l: Value, r: Value) =>
        if ((l ne null) && (r ne null)) l.compareTo(r)
        else if (l ne null) -1
        else if (r ne null) 1
        else 0
      })
      //scalafix:on

      data.iterator
        .takeWhile(_ ne null) // scalafix:ok DisableSyntax.null
        .map(v => (v.value, v.key))
    }

    def register(hash: Int, key: Array[Byte]): this.type = {
      val hsh = if (hash == 0) Fnv1.H else hash
      @tailrec def loop(idx: Int): Unit = {
        val idxHash = hashes(idx)
        if (idxHash == 0) {
          hashes(idx) = hsh
          values(idx) = Value(key)
          length += 1

          if (length > max) ensureCapacity()
        } else if (idxHash != hsh || !values(idx).update(key)) {
          loop((idx + 1) & mask)
        }
      }

      loop(hsh & mask)
      this
    }

    private def ensureCapacity(): Unit = {
      while (length > max) {
        capacity *= 2
        mask = capacity - 1
        max = (LoadFactor * capacity.toFloat).toInt
      }

      val newHashes = new Array[Int](capacity)
      val newValues = new Array[Value](capacity)

      hashes.iterator.zipWithIndex.filter(_._1 != 0).foreach {
        case (hash, i) =>
          var idx = hash & mask
          while (newHashes(idx) != 0) idx = (idx + 1) & mask

          newHashes(idx) = hash
          newValues(idx) = values(i)
      }

      hashes = newHashes
      values = newValues
    }

  }

  object FrequencyDict {

    final class Value(private val bytes: Array[Byte]) extends Comparable[Value] {
      private var counter: Int = 1

      lazy val key: String = new String(bytes, StandardCharsets.UTF_8)
      def value: Int       = counter

      def update(arr: Array[Byte]): Boolean = {
        val same = util.Arrays.equals(bytes, arr)
        if (same) counter += 1
        same
      }

      def compareTo(that: Value): Int = {
        var r = that.counter - this.counter
        var i = 0
        while (r == 0 && i < this.bytes.length && i < that.bytes.length) {
          r = this.bytes(i) - that.bytes(i)
          i += 1
        }

        if (r == 0 && i < that.bytes.length) -1
        else if (r == 0 && i < this.bytes.length) 1
        else r
      }
    }

    object Value {
      def apply(bytes: Array[Byte]): Value =
        new Value(bytes)
    }

    val InitialCapacity = 128
    val LoadFactor      = 0.9f

    def apply(): FrequencyDict = new FrequencyDict(InitialCapacity)

  }

}

File: src/freq01.rs
use std::{
    cmp::Ordering,
    fs::File,
    io::{BufWriter, Read, Write},
};

use clap::Clap;
use memmap::*;

/// FNV1 hash basis
const H: usize = 0x811c_9dc5;
/// FNV1 hash prime
const P: usize = 0x0100_0193;

/// Counts number of unique `[a-zA-Z]+` words in input.
#[derive(Clap, Debug)]
#[clap(version = "0.1.0")]
struct Opts {
    /// Name of input file, or `-` if STDIN should be used.
    input: String,
    /// Name of output file, or `-` if STDOUT should be used.
    output: Option<String>,
}

fn main() {
    let opts: Opts = Opts::parse();
    let input = open_mmap(&opts);

    let mut hash = H;
    let mut dict = FrequencyHashMap::new();

    let mut word_start = 0;
    let mut word_end = 0;
    for (idx, &byte) in input.iter().enumerate() {
        if (b'a' <= byte && byte <= b'z') || (b'A' <= byte && byte <= b'Z') {
            hash ^= (byte & 0x1F) as usize;
            hash *= P;
            word_end = idx + 1;
        } else {
            if word_start < word_end {
                dict.register(hash, &input[word_start..word_end]);
                hash = H;
            }
            word_start = idx + 1;
        }
    }
    if word_start < word_end {
        dict.register(hash, &input[word_start..word_end]);
    }

    let mut output = create_output(&opts);
    for (count, word) in dict.into_iter() {
        writeln!(&mut output, "{} {}", count, word).unwrap_or_else(|e| {
            let output = opts.output.as_ref().map_or("-", |s| s.as_str());
            panic!("Unable to write results in '{}': {}", output, e)
        })
    }
}

struct FrequencyHashMap {
    buckets: Vec<Option<FrequencyHashEntry>>,
    capacity: usize,
    length: usize,
    mask: usize,
    max: usize,
}

#[derive(Clone, Debug)]
struct FrequencyHashEntry {
    key: Box<[u8]>,
    value: usize,
    hash: usize,
}

struct FrequencyHashIntoIter {
    iter: std::vec::IntoIter<Option<FrequencyHashEntry>>,
}

impl FrequencyHashMap {
    const INITIAL: usize = 128;
    const LOAD_FACTOR: f32 = 0.9;

    fn new() -> FrequencyHashMap {
        FrequencyHashMap {
            buckets: vec![None; Self::INITIAL],
            capacity: Self::INITIAL,
            length: 0,
            mask: Self::INITIAL - 1,
            max: (Self::LOAD_FACTOR * Self::INITIAL as f32) as usize,
        }
    }

    fn register(&mut self, hash: usize, word: &[u8]) {
        let mut index = hash & self.mask;
        loop {
            match unsafe { self.buckets.get_unchecked_mut(index) } {
                Some(entry) => {
                    if entry.same_as(hash, word) {
                        entry.value += 1;
                        return;
                    } else {
                        index = (index + 1) & self.mask
                    }
                }
                none => {
                    none.replace(FrequencyHashEntry::new(hash, word));

                    self.length += 1;
                    if self.length > self.max {
                        self.ensure_capacity();
                    }
                    break;
                }
            }
        }
    }

    fn ensure_capacity(&mut self) {
        while self.length > self.max {
            self.capacity *= 2;
            self.mask = self.capacity - 1;
            self.max = (Self::LOAD_FACTOR * self.capacity as f32) as usize;
        }

        let new_buckets = vec![None; self.capacity];
        for bucket in std::mem::replace(&mut self.buckets, new_buckets) {
            if let Some(entry) = bucket {
                let mut index = entry.hash & self.mask;
                loop {
                    match unsafe { self.buckets.get_unchecked_mut(index) } {
                        Some(_) => index = (index + 1) & self.mask,
                        none => {
                            none.replace(entry);
                            break;
                        }
                    }
                }
            }
        }
    }
}

impl IntoIterator for FrequencyHashMap {
    type Item = (usize, String);
    type IntoIter = FrequencyHashIntoIter;

    fn into_iter(self) -> Self::IntoIter {
        let mut buckets = self.buckets;
        buckets.sort_unstable();

        FrequencyHashIntoIter {
            iter: buckets.into_iter(),
        }
    }
}

impl FrequencyHashEntry {
    #[inline]
    fn new(hash: usize, word: &[u8]) -> FrequencyHashEntry {
        FrequencyHashEntry {
            key: word.iter().map(|b| b | 0x20).collect(),
            hash,
            value: 1,
        }
    }

    #[inline]
    fn same_as(&self, hash: usize, word: &[u8]) -> bool {
        hash == self.hash
            && self.key.len() == word.len()
            && Iterator::zip(self.key.iter(), word.iter()).all(|(&l, &r)| l == (r | 0x20))
    }
}

impl Ord for FrequencyHashEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        Ord::cmp(&other.value, &self.value)
            .then_with(|| Ord::cmp(self.key.as_ref(), other.key.as_ref()))
    }
}

impl PartialOrd for FrequencyHashEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(Ord::cmp(&self, &other))
    }
}

impl Eq for FrequencyHashEntry {}

impl PartialEq for FrequencyHashEntry {
    fn eq(&self, other: &Self) -> bool {
        Ord::cmp(&self, &other) == Ordering::Equal
    }
}

impl Iterator for FrequencyHashIntoIter {
    type Item = (usize, String);

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(opt) = self.iter.next() {
            if let Some(entry) = opt {
                let key = std::str::from_utf8(&entry.key).unwrap().to_owned();
                return Some((entry.value, key));
            }
        }
        None
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let (_, upper) = self.iter.size_hint();
        (0, upper)
    }
}

fn open_mmap(opts: &Opts) -> Mmap {
    match opts.input.as_str() {
        "-" => {
            let mut buffer = vec![];
            std::io::stdin()
                .read_to_end(&mut buffer)
                .unwrap_or_else(|e| panic!("Unable to read STDIN: {}", e));
            let mut mmap = MmapOptions::new()
                .len(buffer.len())
                .map_anon()
                .unwrap_or_else(|e| panic!("Unable to read STDIN: {}", e));
            mmap.copy_from_slice(&buffer);
            mmap.make_read_only()
                .unwrap_or_else(|e| panic!("Unable to read STDIN: {}", e))
        }
        fnm => {
            let file = File::open(fnm)
                .unwrap_or_else(|e| panic!("Unable to open '{}' for reading: {}", fnm, e));
            unsafe {
                MmapOptions::new()
                    .map(&file)
                    .unwrap_or_else(|e| panic!("Unable to read '{}' in memory: {}", fnm, e))
            }
        }
    }
}

fn create_output(opts: &Opts) -> Box<dyn Write> {
    match opts.output.as_deref() {
        Some("-") | None => Box::new(BufWriter::new(std::io::stdout())),
        Some(fnm) => {
            let file = File::create(fnm)
                .unwrap_or_else(|e| panic!("Unable to open '{}' for writing: {}", fnm, e));
            Box::new(BufWriter::new(file))
        }
    }
}
